pkg/api/kubegraph/analysis/hpa.go

```diff
@@ -0,0 +1,177 @@
+	// Currently, the only supported scale metric is CPU utilization, so without this metric an HPA is useless.
+	HPAMissingCPUTargetError = "HPAMissingCPUTarget"
+	// HPAOverlappingScaleRefWarning denotes a warning where a Horizontal Pod Autoscaler scales an object that is scaled by some other object as well
+	HPAOverlappingScaleRefWarning = "HPAOverlappingScaleRef"
+)
+
+// FindHPASpecsMissingCPUTargets scans the graph in search of HorizontalPodAutoscalers that are missing a CPU utilization target.
+// As of right now, the only metric that HPAs can use to scale pods is the CPU utilization, so if a HPA is missing this target it
+// is effectively useless.
+func FindHPASpecsMissingCPUTargets(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+	for _, uncastNode := range graph.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		hpa, ok := node.Object().(*extensions.HorizontalPodAutoscaler)
+		if !ok {
+			continue
+		}
+
+		if hpa.Spec.CPUUtilization == nil {
+			markers = append(markers, osgraph.Marker{
+				Node:     node,
+				Severity: osgraph.ErrorSeverity,
+				Key:      HPAMissingCPUTargetError,
+				Message:  fmt.Sprintf("%s is missing a CPU utilization target", namer.ResourceName(node)),
+			})
+		}
+	}
+
+	return markers
+}
+
+// FindHPASpecsMissingScaleRefs finds all Horizontal Pod Autoscalers whose scale reference points to an object that doesn't exist
+// or that the client does not have the permission to see.
+func FindHPASpecsMissingScaleRefs(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+	for _, uncastNode := range graph.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		scaledObjects := graph.SuccessorNodesByEdgeKind(
+			uncastNode,
+			kubegraph.ScalingEdgeKind,
+		)
+
+		if len(scaledObjects) < 1 {
+			markers = append(markers, createMissingScaleRefMarker(node, nil, namer))
+		}
+
+		for _, scaleRef := range scaledObjects {
+			if existenceChecker, ok := scaleRef.(osgraph.ExistenceChecker); ok && !existenceChecker.Found() {
+				// if this node is synthetic, we can't be sure that the HPA is scaling something that actually exists
+				markers = append(markers, createMissingScaleRefMarker(node, scaleRef, namer))
+			}
+		}
+	}
+
+	return markers
+}
+
+func createMissingScaleRefMarker(hpaNode *kubenodes.HorizontalPodAutoscalerNode, scaleRef graphapi.Node, namer osgraph.Namer) osgraph.Marker {
+	return osgraph.Marker{
+		Node:         hpaNode,
+		Severity:     osgraph.ErrorSeverity,
+		RelatedNodes: []graphapi.Node{scaleRef},
+		Key:          HPAMissingScaleRefError,
+		Message:      fmt.Sprintf("%s is attempting to scale %s/%s, which doesn't exist", namer.ResourceName(hpaNode), hpaNode.Spec.ScaleRef.Kind, hpaNode.Spec.ScaleRef.Name),
```

Comment:
I don't know if dangling autoscalers are any good. They most probably have been created with `oc autoscale` or via a template so the autoscaled object has always pre-existed. What worries me is accidental adoption.